tensorflow/compiler/mlir/xla/ir/infer_fusibility_op_interface.td

```diff
@@ -0,0 +1,123 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This file contains inferFusiblityOpInterface, which is used to guide
+// fusion decsion.
+
+#ifndef MLIR_INFER_FUSIBILITY_OP_INTERFACE
+#define MLIR_INFER_FUSIBILITY_OP_INTERFACE
+
+include "mlir/IR/OpBase.td"
+
+// OpInterface to query if an op is fusible and to query the shape equality
+// constraint among the inputs and outputs of an op.
+def InferFusibilityOpInterface : OpInterface<"InferFusibilityOpInterface"> {
+  let description = [{
+    Interface to query if an op is fusible and to query the shape equality
+    constraint among the inputs and outputs of an op.
+  }];
+
+  let methods = [
+    InterfaceMethod<
+      /*desc=*/[{If true, this op can be fused with its operands
+      }],
+      /*retTy=*/"bool",
+      /*methodName=*/"isFusibleWithOperand",
+      /*args=*/(ins),
+      /*methodBody=*/[{}],
+      /*defaultImplementation=*/[{
+        /// Returns whether this op can be fused with its operands
+        return true;
+      }]
+    >,
+    InterfaceMethod<
+      /*desc=*/[{If true, this op can be fused with its consumers
+      }],
+      /*retTy=*/"bool",
+      /*methodName=*/"isFusibleWithConsumer",
+      /*args=*/(ins),
+      /*methodBody=*/[{}],
+      /*defaultImplementation=*/[{
+        /// Return whether this op can be fused withh its consumers
+        return true;
+      }]
+    >,
+    InterfaceMethod<
+      /*desc=*/"Return whether two inputs have the same shape (assuming no"
+               "implicit broadcasting).",
+      /*retTy=*/"bool",
+      /*methodName=*/"inferInputsShapeEquality",
```

Comment:
the int type input of this function here is the index of op's inputs or outputs.

Reply:
@wyzero why can't the default implementation look for the actual shapes of the inputs instead of just if these refer to the same input?

```
/*defaultImplementation=*/[{
  Operation *op = this->getOperation();
  assert(lhs < op->getNumOperands() && lhs >= 0 &&
                rhs < op->getNumOperands() && rhs >= 0);
  if (lhs == rhs) return true;
  Type lhs_ty = op->getOperand(lhs).getType();
  Type rhs_ty = op->getOperand(rhs).getType();
  // ... check for shape equality
```